File: web-sdks/dart/lib/src/core/response.dart
/// Enhanced response system for GroupVAN SDK
///
/// Provides comprehensive response handling with metadata, caching,
/// and type-safe data access following Flutter/Dart best practices.
library response;

/// Enhanced response wrapper with comprehensive metadata
class GroupVanResponse<T> {
  /// The actual response data
  final T data;

  /// HTTP status code
  final int statusCode;

  /// Response headers
  final Map<String, String> headers;

  /// Request metadata
  final RequestMetadata requestMetadata;

  /// Response metadata
  final ResponseMetadata responseMetadata;

  /// Session ID if present
  final String? sessionId;

  /// Whether the response came from cache
  final bool fromCache;

  /// Cache timestamp if applicable
  final DateTime? cacheTimestamp;

  GroupVanResponse({
    required this.data,
    required this.statusCode,
    required this.headers,
    required this.requestMetadata,
    required this.responseMetadata,
    this.sessionId,
    this.fromCache = false,
    this.cacheTimestamp,
  });

  /// Whether the response was successful (2xx status code)
  bool get isSuccessful => statusCode >= 200 && statusCode < 300;

  /// Whether the response was from cache
  bool get isCached => fromCache;

  /// Response age if cached
  Duration? get cacheAge {
    if (cacheTimestamp == null) return null;
    return DateTime.now().difference(cacheTimestamp!);
  }

  /// Create a copy with different data
  GroupVanResponse<U> copyWith<U>({
    U? data,
    int? statusCode,
    Map<String, String>? headers,
    RequestMetadata? requestMetadata,
    ResponseMetadata? responseMetadata,
    String? sessionId,
    bool? fromCache,
    DateTime? cacheTimestamp,
  }) {
    return GroupVanResponse<U>(
      data: data ?? this.data as U,
      statusCode: statusCode ?? this.statusCode,
      headers: headers ?? this.headers,
      requestMetadata: requestMetadata ?? this.requestMetadata,
      responseMetadata: responseMetadata ?? this.responseMetadata,
      sessionId: sessionId ?? this.sessionId,
      fromCache: fromCache ?? this.fromCache,
      cacheTimestamp: cacheTimestamp ?? this.cacheTimestamp,
    );
  }

  @override
  String toString() {
    return 'GroupVanResponse<$T>(status: $statusCode, cached: $fromCache, session: $sessionId)';
  }
}

/// Request metadata for debugging and monitoring
class RequestMetadata {
  /// HTTP method used
  final String method;

  /// Full request URL
  final String url;

  /// Request headers sent
  final Map<String, String> headers;

  /// Request body if applicable
  final dynamic body;

  /// Request timestamp
  final DateTime timestamp;

  /// Request timeout duration
  final Duration timeout;

  /// Retry attempt number (0 for first attempt)
  final int retryAttempt;

  /// Request correlation ID for tracing
  final String correlationId;

  RequestMetadata({
    required this.method,
    required this.url,
    required this.headers,
    this.body,
    required this.timestamp,
    required this.timeout,
    this.retryAttempt = 0,
    required this.correlationId,
  });

  @override
  String toString() {
    return 'RequestMetadata(method: $method, url: $url, retry: $retryAttempt, correlation: $correlationId)';
  }
}

/// Response metadata for debugging and monitoring
class ResponseMetadata {
  /// Response timestamp
  final DateTime timestamp;

  /// Response processing duration
  final Duration duration;

  /// Response size in bytes
  final int? sizeBytes;

  /// Whether response was gzipped
  final bool compressed;

  /// Response encoding
  final String? encoding;

  /// Server information
  final String? server;

  /// Response correlation ID
  final String correlationId;

  ResponseMetadata({
    required this.timestamp,
    required this.duration,
    this.sizeBytes,
    this.compressed = false,
    this.encoding,
    this.server,
    required this.correlationId,
  });

  @override
  String toString() {
    return 'ResponseMetadata(duration: ${duration.inMilliseconds}ms, size: ${sizeBytes ?? 'unknown'} bytes, correlation: $correlationId)';
  }
}

/// Paginated response wrapper
class PaginatedResponse<T> extends GroupVanResponse<List<T>> {
  /// Current page number
  final int page;

  /// Items per page
  final int limit;

  /// Total number of items
  final int totalCount;

  /// Total number of pages
  final int totalPages;

  /// Whether there are more pages
  final bool hasNextPage;

  /// Whether there are previous pages
  final bool hasPreviousPage;

  PaginatedResponse({
    required super.data,
    required super.statusCode,
    required super.headers,
    required super.requestMetadata,
    required super.responseMetadata,
    super.sessionId,
    super.fromCache,
    super.cacheTimestamp,
    required this.page,
    required this.limit,
    required this.totalCount,
  }) : totalPages = (totalCount / limit).ceil(),
       hasNextPage = (page * limit) < totalCount,
       hasPreviousPage = page > 1;

  @override
  String toString() {
    return 'PaginatedResponse<$T>(status: $statusCode, page: $page/$totalPages, items: ${data.length}/$totalCount)';
  }
}

/// Response builder for creating responses with proper metadata
class ResponseBuilder<T> {
  T? _data;
  int? _statusCode;
  Map<String, String>? _headers;
  RequestMetadata? _requestMetadata;
  ResponseMetadata? _responseMetadata;
  String? _sessionId;
  bool _fromCache = false;
  DateTime? _cacheTimestamp;

  ResponseBuilder<T> data(T data) {
    _data = data;
    return this;
  }

  ResponseBuilder<T> statusCode(int statusCode) {
    _statusCode = statusCode;
    return this;
  }

  ResponseBuilder<T> headers(Map<String, String> headers) {
    _headers = headers;
    return this;
  }

  ResponseBuilder<T> requestMetadata(RequestMetadata metadata) {
    _requestMetadata = metadata;
    return this;
  }

  ResponseBuilder<T> responseMetadata(ResponseMetadata metadata) {
    _responseMetadata = metadata;
    return this;
  }

  ResponseBuilder<T> sessionId(String? sessionId) {
    _sessionId = sessionId;
    return this;
  }

  ResponseBuilder<T> fromCache(bool cached, [DateTime? timestamp]) {
    _fromCache = cached;
    _cacheTimestamp = timestamp;
    return this;
  }

  GroupVanResponse<T> build() {
    if (_data == null ||
        _statusCode == null ||
        _headers == null ||
        _requestMetadata == null ||
        _responseMetadata == null) {
      throw StateError('Missing required response data');
    }

    return GroupVanResponse<T>(
      data: _data as T,
      statusCode: _statusCode!,
      headers: _headers!,
      requestMetadata: _requestMetadata!,
      responseMetadata: _responseMetadata!,
      sessionId: _sessionId,
      fromCache: _fromCache,
      cacheTimestamp: _cacheTimestamp,
    );
  }
}

/// Result type for operations that may fail
sealed class Result<T> {
  const Result();

  /// Whether this result is successful
  bool get isSuccess => this is Success<T>;

  /// Whether this result is a failure
  bool get isFailure => this is Failure<T>;

  /// Get the success value (throws if failure)
  T get value {
    return switch (this) {
      Success<T>(value: final value) => value,
      Failure<T>() => throw StateError('Cannot get value from failure result'),
    };
  }

  /// Get the error (throws if success)
  Exception get error {
    return switch (this) {
      Success<T>() => throw StateError('Cannot get error from success result'),
      Failure<T>(error: final error) => error,
    };
  }

  /// Transform success value
  Result<U> map<U>(U Function(T value) transform) {
    return switch (this) {
      Success<T>(value: final value) => Success(transform(value)),
      Failure<T>(error: final error) => Failure(error),
    };
  }

  /// Transform error
  Result<T> mapError(Exception Function(Exception error) transform) {
    return switch (this) {
      Success<T>(value: final value) => Success(value),
      Failure<T>(error: final error) => Failure(transform(error)),
    };
  }

  /// Handle both success and failure cases
  U fold<U>(
    U Function(Exception error) onFailure,
    U Function(T value) onSuccess,
  ) {
    return switch (this) {
      Success<T>(value: final value) => onSuccess(value),
      Failure<T>(error: final error) => onFailure(error),
    };
  }
}

/// Success result
final class Success<T> extends Result<T> {
  @override
  final T value;

  const Success(this.value);

  @override
  String toString() => 'Success($value)';

  @override
  bool operator ==(Object other) {
    return identical(this, other) ||
        other is Success<T> && value == other.value;
  }

  @override
  int get hashCode => value.hashCode;
}

/// Failure result
final class Failure<T> extends Result<T> {
  @override
  final Exception error;

  const Failure(this.error);

  @override
  String toString() => 'Failure($error)';

  @override
  bool operator ==(Object other) {
    return identical(this, other) ||
        other is Failure<T> && error == other.error;
  }

  @override
  int get hashCode => error.hashCode;
}

File: web-sdks/javascript/src/core/response.js
/**
 * Enhanced response system for GroupVAN SDK
 *
 * Provides comprehensive response handling with metadata, caching,
 * and type-safe data access.
 */

/**
 * Request metadata for debugging and monitoring
 */
export class RequestMetadata {
  /**
   * @param {Object} options
   * @param {string} options.method - HTTP method used
   * @param {string} options.url - Full request URL
   * @param {Object} options.headers - Request headers sent
   * @param {*} [options.body] - Request body if applicable
   * @param {Date} options.timestamp - Request timestamp
   * @param {number} options.timeout - Request timeout in milliseconds
   * @param {number} [options.retryAttempt=0] - Retry attempt number
   * @param {string} options.correlationId - Request correlation ID for tracing
   */
  constructor({ method, url, headers, body = null, timestamp, timeout, retryAttempt = 0, correlationId }) {
    this.method = method;
    this.url = url;
    this.headers = headers;
    this.body = body;
    this.timestamp = timestamp;
    this.timeout = timeout;
    this.retryAttempt = retryAttempt;
    this.correlationId = correlationId;
  }

  toString() {
    return `RequestMetadata(method: ${this.method}, url: ${this.url}, retry: ${this.retryAttempt}, correlation: ${this.correlationId})`;
  }
}

/**
 * Response metadata for debugging and monitoring
 */
export class ResponseMetadata {
  /**
   * @param {Object} options
   * @param {Date} options.timestamp - Response timestamp
   * @param {number} options.durationMs - Response processing duration in milliseconds
   * @param {number} [options.sizeBytes] - Response size in bytes
   * @param {boolean} [options.compressed=false] - Whether response was gzipped
   * @param {string} [options.encoding] - Response encoding
   * @param {string} [options.server] - Server information
   * @param {string} options.correlationId - Response correlation ID
   */
  constructor({ timestamp, durationMs, sizeBytes = null, compressed = false, encoding = null, server = null, correlationId }) {
    this.timestamp = timestamp;
    this.durationMs = durationMs;
    this.sizeBytes = sizeBytes;
    this.compressed = compressed;
    this.encoding = encoding;
    this.server = server;
    this.correlationId = correlationId;
  }

  toString() {
    return `ResponseMetadata(duration: ${this.durationMs}ms, size: ${this.sizeBytes ?? 'unknown'} bytes, correlation: ${this.correlationId})`;
  }
}

/**
 * Enhanced response wrapper with comprehensive metadata
 * @template T
 */
export class GroupVanResponse {
  /**
   * @param {Object} options
   * @param {T} options.data - The actual response data
   * @param {number} options.statusCode - HTTP status code
   * @param {Object} options.headers - Response headers
   * @param {RequestMetadata} options.requestMetadata - Request metadata
   * @param {ResponseMetadata} options.responseMetadata - Response metadata
   * @param {boolean} [options.fromCache=false] - Whether the response came from cache
   * @param {Date} [options.cacheTimestamp] - Cache timestamp if applicable
   */
  constructor({ data, statusCode, headers, requestMetadata, responseMetadata, fromCache = false, cacheTimestamp = null }) {
    this.data = data;
    this.statusCode = statusCode;
    this.headers = headers;
    this.requestMetadata = requestMetadata;
    this.responseMetadata = responseMetadata;
    this.fromCache = fromCache;
    this.cacheTimestamp = cacheTimestamp;
  }

  /**
   * Whether the response was successful (2xx status code)
   * @returns {boolean}
   */
  get isSuccessful() {
    return this.statusCode >= 200 && this.statusCode < 300;
  }

  /**
   * Whether the response was from cache
   * @returns {boolean}
   */
  get isCached() {
    return this.fromCache;
  }

  /**
   * Response age if cached (in milliseconds)
   * @returns {number|null}
   */
  get cacheAgeMs() {
    if (!this.cacheTimestamp) return null;
    return Date.now() - this.cacheTimestamp.getTime();
  }

  toString() {
    return `GroupVanResponse(status: ${this.statusCode}, cached: ${this.fromCache})`;
  }
}

/**
 * Paginated response wrapper
 * @template T
 * @extends {GroupVanResponse<T[]>}
 */
export class PaginatedResponse extends GroupVanResponse {
  /**
   * @param {Object} options
   * @param {T[]} options.data - The response data array
   * @param {number} options.statusCode - HTTP status code
   * @param {Object} options.headers - Response headers
   * @param {RequestMetadata} options.requestMetadata - Request metadata
   * @param {ResponseMetadata} options.responseMetadata - Response metadata
   * @param {number} options.page - Current page number
   * @param {number} options.limit - Items per page
   * @param {number} options.totalCount - Total number of items
   * @param {boolean} [options.fromCache=false] - Whether from cache
   * @param {Date} [options.cacheTimestamp] - Cache timestamp
   */
  constructor({ data, statusCode, headers, requestMetadata, responseMetadata, page, limit, totalCount, fromCache = false, cacheTimestamp = null }) {
    super({ data, statusCode, headers, requestMetadata, responseMetadata, fromCache, cacheTimestamp });
    this.page = page;
    this.limit = limit;
    this.totalCount = totalCount;
    this.totalPages = Math.ceil(totalCount / limit);
    this.hasNextPage = (page * limit) < totalCount;
    this.hasPreviousPage = page > 1;
  }

  toString() {
    return `PaginatedResponse(status: ${this.statusCode}, page: ${this.page}/${this.totalPages}, items: ${this.data.length}/${this.totalCount})`;
  }
}

/**
 * Response builder for creating responses with proper metadata
 * @template T
 */
export class ResponseBuilder {
  constructor() {
    this._data = null;
    this._statusCode = null;
    this._headers = null;
    this._requestMetadata = null;
    this._responseMetadata = null;
    this._fromCache = false;
    this._cacheTimestamp = null;
  }

  /**
   * @param {T} data
   * @returns {ResponseBuilder<T>}
   */
  data(data) {
    this._data = data;
    return this;
  }

  /**
   * @param {number} statusCode
   * @returns {ResponseBuilder<T>}
   */
  statusCode(statusCode) {
    this._statusCode = statusCode;
    return this;
  }

  /**
   * @param {Object} headers
   * @returns {ResponseBuilder<T>}
   */
  headers(headers) {
    this._headers = headers;
    return this;
  }

  /**
   * @param {RequestMetadata} metadata
   * @returns {ResponseBuilder<T>}
   */
  requestMetadata(metadata) {
    this._requestMetadata = metadata;
    return this;
  }

  /**
   * @param {ResponseMetadata} metadata
   * @returns {ResponseBuilder<T>}
   */
  responseMetadata(metadata) {
    this._responseMetadata = metadata;
    return this;
  }

  /**
   * @param {boolean} cached
   * @param {Date} [timestamp]
   * @returns {ResponseBuilder<T>}
   */
  fromCache(cached, timestamp = null) {
    this._fromCache = cached;
    this._cacheTimestamp = timestamp;
    return this;
  }

  /**
   * @returns {GroupVanResponse<T>}
   * @throws {Error} If required data is missing
   */
  build() {
    if (
      this._data === null ||
      this._statusCode === null ||
      this._headers === null ||
      this._requestMetadata === null ||
      this._responseMetadata === null
    ) {
      throw new Error('Missing required response data');
    }

    return new GroupVanResponse({
      data: this._data,
      statusCode: this._statusCode,
      headers: this._headers,
      requestMetadata: this._requestMetadata,
      responseMetadata: this._responseMetadata,
      fromCache: this._fromCache,
      cacheTimestamp: this._cacheTimestamp,
    });
  }
}

/**
 * Result type for operations that may fail
 * @template T
 */
export class Result {
  /**
   * @param {boolean} isSuccess
   * @param {T} [value]
   * @param {Error} [error]
   */
  constructor(isSuccess, value = null, error = null) {
    this._isSuccess = isSuccess;
    this._value = value;
    this._error = error;
  }

  /**
   * Create a success result
   * @template T
   * @param {T} value
   * @returns {Result<T>}
   */
  static success(value) {
    return new Result(true, value, null);
  }

  /**
   * Create a failure result
   * @template T
   * @param {Error} error
   * @returns {Result<T>}
   */
  static failure(error) {
    return new Result(false, null, error);
  }

  /**
   * Whether this result is successful
   * @returns {boolean}
   */
  get isSuccess() {
    return this._isSuccess;
  }

  /**
   * Whether this result is a failure
   * @returns {boolean}
   */
  get isFailure() {
    return !this._isSuccess;
  }

  /**
   * Get the success value (throws if failure)
   * @returns {T}
   * @throws {Error} If result is a failure
   */
  get value() {
    if (!this._isSuccess) {
      throw new Error('Cannot get value from failure result');
    }
    return this._value;
  }

  /**
   * Get the error (throws if success)
   * @returns {Error}
   * @throws {Error} If result is a success
   */
  get error() {
    if (this._isSuccess) {
      throw new Error('Cannot get error from success result');
    }
    return this._error;
  }

  /**
   * Get the value or null if failure
   * @returns {T|null}
   */
  getOrNull() {
    return this._isSuccess ? this._value : null;
  }

  /**
   * Get the value or a default if failure
   * @param {T} defaultValue
   * @returns {T}
   */
  getOrDefault(defaultValue) {
    return this._isSuccess ? this._value : defaultValue;
  }

  /**
   * Transform success value
   * @template U
   * @param {function(T): U} transform
   * @returns {Result<U>}
   */
  map(transform) {
    if (this._isSuccess) {
      return Result.success(transform(this._value));
    }
    return Result.failure(this._error);
  }

  /**
   * Transform error
   * @param {function(Error): Error} transform
   * @returns {Result<T>}
   */
  mapError(transform) {
    if (!this._isSuccess) {
      return Result.failure(transform(this._error));
    }
    return Result.success(this._value);
  }

  /**
   * Handle both success and failure cases
   * @template U
   * @param {function(Error): U} onFailure
   * @param {function(T): U} onSuccess
   * @returns {U}
   */
  fold(onFailure, onSuccess) {
    if (this._isSuccess) {
      return onSuccess(this._value);
    }
    return onFailure(this._error);
  }

  toString() {
    if (this._isSuccess) {
      return `Success(${this._value})`;
    }
    return `Failure(${this._error})`;
  }
}
